crates/sparrow-journal/src/batch_io.rs

```diff
@@ -0,0 +1,216 @@
+    dictionaries_by_id: HashMap<i64, ArrayRef>,
+}
+
+impl BatchDecoder {
+    pub fn new(schema: SchemaRef) -> Self {
+        Self {
+            schema,
+            dictionaries_by_id: HashMap::new(),
+        }
+    }
+
+    /// Decode the batch from the given entry.
+    ///
+    /// If it was completely written, will return `Some(batch)`. If it wasn't completely
+    /// written, will return `None`.
+    pub fn decode(
+        &mut self,
+        entry: &mut okaywal::Entry<'_>,
+    ) -> error_stack::Result<Option<Batch>, Error> {
+        // Read the chunks.
+        let Some(chunks) = entry
+            .read_all_chunks()
+            .into_report()
+            .change_context(Error::ReadBatch)?
+        else {
+            return Ok(None);
+        };
+
+        let mut chunks = chunks.into_iter();
+
+        let up_to_time = chunks.next().expect("at least one entry");
+        let up_to_time = RowTime::from_bytes(up_to_time.try_into().expect("row time"));
+
+        let mut chunk_pairs = chunks.tuples().peekable();
+        while let Some((metadata, batch)) = chunk_pairs.next() {
+            let message = decode_message(&metadata)?;
+            error_stack::ensure!(
+                message.bodyLength() as usize == batch.len(),
+                Error::ReadBatch
+            );
+
+            let buf = Buffer::from(batch);
+
+            if chunk_pairs.peek().is_some() {
+                error_stack::ensure!(
+                    message.header_type() == arrow_ipc::MessageHeader::DictionaryBatch,
+                    Error::ReadBatch
+                );
+
+                let batch = message.header_as_dictionary_batch().unwrap();
+                arrow_ipc::reader::read_dictionary(
+                    &buf,
+                    batch,
+                    &self.schema,
+                    &mut self.dictionaries_by_id,
+                    &message.version(),
+                )
+                .into_report()
+                .change_context(Error::ReadBatch)?;
+            } else {
+                error_stack::ensure!(
+                    message.header_type() == arrow_ipc::MessageHeader::RecordBatch,
+                    Error::ReadBatch
+                );
+                let batch = message.header_as_record_batch().unwrap();
+                let batch = arrow_ipc::reader::read_record_batch(
+                    &buf,
+                    batch,
+                    self.schema.clone(),
+                    &self.dictionaries_by_id,
+                    None,
+                    &message.version(),
+                )
+                .into_report()
+                .change_context(Error::ReadBatch)?;
+
+                return Ok(Some(Batch::new_with_data(
+                    batch.column(3).clone(),
+                    batch.column(0).clone(),
+                    batch.column(1).clone(),
+                    batch.column(2).clone(),
+                    up_to_time,
+                )));
+            }
+        }
```

Comment:
This logic feels interesting. Iterating through all chunk pairs and adding the batch to the buffer, but once we peek and see we're at the last chunk pair, return a record batch? 

Is it done this way because we have to set the last batch's message with the `header_as_record_batch()` instead of `as_dict_batch()`?

Reply:
Yeah -- this roughly based on what the Arrow IPC writer does under the hood. Each "dictionary column" becomes a dictionary batch, and then you have the final batch. So you end up with:

- Dict 1 Metadata, Dict 1 Data
- Dict 2 Metadata, Dict 2 Data
- ...
- Dict N Metadata, Dict N Data
- Batch Metadata, Batch Metadata

In many cases, there will be no dicts, but figured it was worth keeping.